### DISCUSSION TITLE
RLS setup issue

### DISCUSSION THREAD

**Discussion by jaynguyens** [supabase/supabase#2911] [Questions] [answered]
What I am trying to do is fetch data for authenticated user that they created. So after user logged in, they can see only the data they created. 

Currently I have a `sessions` table:
![image](https://user-images.githubusercontent.com/74240381/130226210-1381c658-b48c-44dd-8d15-6be9c3c60370.png)

and in authentication table:
![image](https://user-images.githubusercontent.com/74240381/130226455-f9c5534b-cb7d-4040-9262-033f5b83e234.png)

with the policy: 
```sql
ALTER POLICY "Enable all for users based on user_id" ON public.sessions FOR ALL
USING (auth.uid() = user_id)
WITH CHECK (auth.uid() = user_id);
```
and ofc fetch the database
```javascript
const getResponse = await supabase
  .from('sessions')
  .select('*')
```
However, it doesn't return any data back. I'm not sure what I missed from the docs. I currently have a assumption because of the`user_id` column is a foreign key, somehow, it the auth.uid() which is uuid can't filters against the foreign key? but this just seen odd to me...

Do I need to join two table and then make a check? or is there something that I miss?

Any help would be greatly appreciated, thanks 🙏

--- 
After a bit of testing, `using (user_id = '...')` working correct, but the `auth.uid()` not. So something isn't correct with the `auth.uid()`

**Comment by dshukertjr** [comment, marked as answer]
Hi @jaynguyens!

Hmm, your setup seems correct to me. Could you confirmed that you have a signed in user right before you fetch the data from your sessions table?

```typescript
const user = supabase.auth.user()
console.log('user', user) // Check if you are logged in for sure here

const getResponse = await supabase
  .from('sessions')
  .select('*')
```

**Reply by jaynguyens** [reply]
I forgot to mention am using nextjs and this fetch function is happen on the api route. To get the user I use `const { user } = await supabase.auth.api.getUserByCookie(req)` but not straight from the `supabase.auth.user()`.

This is the code
```js
const Sessions = async (req: NextApiRequest, res: NextApiResponse) => {
    const { body: { userId }, method } = req
    const { user } = await supabase.auth.api.getUserByCookie(req)

    switch (method) {

        case 'GET':
            const getResponse = await supabase
                .from('sessions')
                .select('*')
                .eq('user_id', user?.id)

            res.status(200).json(getResponse)
            break

        default:
            res.setHeader('Allow', ['GET', 'POST'])
            res.status(405).end(`Method ${method} Not Allowed`)
    }
}

```

**Reply by dshukertjr** [reply]
@jaynguyens 
I see! That should be the cause!

If you want to make a request from server as a signed in user, calling `getUserByCookie` isn't enough. You can read the discussion [here](https://github.com/supabase/supabase/discussions/1094) where they tackle similar problem, but in short, you need to do something like this on your API route:
```typescript
const supabase = createClient(url, anonKey);
supabase.auth.setAuth(jwt);

// After you call `setAuth()`, you can make request to Supabase as the signed in user!
const getResponse = await supabase
  .from('sessions')
  .select('*')
  .eq('user_id', user?.id)
```

Where `jwt` is sent from the client as request parameter. This is how you can obtain it on the client. 

```typescript
const { user, session, error } = await supabase
    .auth.signIn({ email: 'ant@boo.com', password: 'smoothmoves11'})
const jwt = session.access_token;
```

**Reply by jaynguyens** [reply]
Great it works, thanks you @dshukertjr. I going to adda bit more information to your answer, for future references someone got the same issue.

We need to add additional header to the http requests, am using `ky` but it would work the same for `axios` or `fetch`.

```javascript
const api = ky.extend({
    hooks: {
        beforeRequest: [
            (request) => {
                const session = supabase.auth.session()
                const token = session?.access_token

                request.headers.set('Authorization', `Bearer ${token}`);
            }
        ]
    }
});
```

And on api route, 
```js
const token = req.cookies["sb:token"]
supabase.auth.setAuth(token)
```

**Reply by minimaluminium** [reply]
Ohhh, this exact issue was driving me insane. Even though the RLS was configured correctly, the `insert()` wasn't working. Had no idea I actually had to set authentication when a request was made in an API route. Thanks a lot!

**Comment by markwcollins** [comment]
I faced the same issue. When I did
`create policy "Can create outfits." on outfits for insert with check (auth.uid() = user_id);`

I get the same error message  "message": "new row violates row-level security policy for table \"outfits\"",

Then when I went to the policy UI I saw that the above policy had been created as `(uid() = user_id)`. I then manually changed it to `auth.uid() = user_id`, it then seemed to work.

**Reply by jaynguyens** [reply]
For some reasons, it doesn't let me change to `auth.uid() = user_id` - it keep convert back to `uid() = user_id`

**Reply by markwcollins** [reply]
Yeah the same thing is happening to me now too. Ignore my previous answer. I'm still getting an error

**Comment by markwcollins** [comment]
@jaynguyens looks like you are using a uuid as the primary key. Are you creating the uuid in the DB or passing it in with the API request? I've noticed some strange behaviours

**Reply by jaynguyens** [reply]
I don't think it matter, but mine is auto generated primary key
